### ISSUE TITLE
`error: The name `set_advertised_protocols' does not exist in the context of `GLib.TlsConnection?'`

### ISSUE THREAD

**Issue by paulmenzel** [dino/dino#932]
Building v0.1.0-152-g5954f77, the current repository HEAD, with GLib 2.64.1 and Vala 0.36.15, the build fails with the error below.

```
[128/769] Generating C code for target CRYPTO_VALA_C
/dev/shm/dino/plugins/crypto-vala/vapi/gcrypt.vapi:503.3-503.7: warning: argument `lower_case_cname' never used
        [CCode (lower_case_cname = "cipher_")]
         ^^^^^
Compilation succeeded - 1 warning(s)
[130/769] Generating C code for target ENGINE_VALA_C
FAILED: xmpp-vala/src/core/namespace_state.c xmpp-vala/src/core/stanza_attribute.c xmpp-vala/src/core/stanza_node.c xmpp-vala/src/core/stanza_reader.c xmpp-vala/src/core/stanza_writer.c xmpp-vala/src/core/xmpp_log.c xmpp-vala/src/core/xmpp_stream.c xmpp-vala/src/module/bind.c xmpp-vala/src/module/bookmarks_provider.c xmpp-vala/src/module/conference.c xmpp-vala/src/module/iq/module.c xmpp-vala/src/module/iq/stanza.c xmpp-vala/src/module/jid.c xmpp-vala/src/module/message/module.c xmpp-vala/src/module/message/stanza.c xmpp-vala/src/module/presence/flag.c xmpp-vala/src/module/presence/module.c xmpp-vala/src/module/presence/stanza.c xmpp-vala/src/module/roster/flag.c xmpp-vala/src/module/roster/item.c xmpp-vala/src/module/roster/module.c xmpp-vala/src/module/roster/versioning_module.c xmpp-vala/src/module/sasl.c xmpp-vala/src/module/session.c xmpp-vala/src/module/stanza.c xmpp-vala/src/module/stanza_error.c xmpp-vala/src/module/stream_error.c xmpp-vala/src/module/tls.c xmpp-vala/src/module/util.c xmpp-vala/src/module/xep/0048_bookmarks.c xmpp-vala/src/module/xep/0048_conference.c xmpp-vala/src/module/xep/0402_bookmarks2.c xmpp-vala/src/module/xep/0004_data_forms.c xmpp-vala/src/module/xep/0030_service_discovery/flag.c xmpp-vala/src/module/xep/0030_service_discovery/identity.c xmpp-vala/src/module/xep/0030_service_discovery/info_result.c xmpp-vala/src/module/xep/0030_service_discovery/item.c xmpp-vala/src/module/xep/0030_service_discovery/items_result.c xmpp-vala/src/module/xep/0030_service_discovery/module.c xmpp-vala/src/module/xep/0045_muc/flag.c xmpp-vala/src/module/xep/0045_muc/module.c xmpp-vala/src/module/xep/0045_muc/status_code.c xmpp-vala/src/module/xep/0047_in_band_bytestreams.c xmpp-vala/src/module/xep/0049_private_xml_storage.c xmpp-vala/src/module/xep/0054_vcard/module.c xmpp-vala/src/module/xep/0060_pubsub.c xmpp-vala/src/module/xep/0065_socks5_bytestreams.c xmpp-vala/src/module/xep/0066_out_of_band_data.c xmpp-vala/src/module/xep/0077_in_band_registration.c xmpp-vala/src/module/xep/0082_date_time_profiles.c xmpp-vala/src/module/xep/0084_user_avatars.c xmpp-vala/src/module/xep/0085_chat_state_notifications.c xmpp-vala/src/module/xep/0115_entity_capabilities.c xmpp-vala/src/module/xep/0166_jingle.c xmpp-vala/src/module/xep/0184_message_delivery_receipts.c xmpp-vala/src/module/xep/0191_blocking_command.c xmpp-vala/src/module/xep/0198_stream_management.c xmpp-vala/src/module/xep/0199_ping.c xmpp-vala/src/module/xep/0203_delayed_delivery.c xmpp-vala/src/module/xep/0234_jingle_file_transfer.c xmpp-vala/src/module/xep/0260_jingle_socks5_bytestreams.c xmpp-vala/src/module/xep/0261_jingle_in_band_bytestreams.c xmpp-vala/src/module/xep/0280_message_carbons.c xmpp-vala/src/module/xep/0308_last_message_correction.c xmpp-vala/src/module/xep/0313_message_archive_management.c xmpp-vala/src/module/xep/0333_chat_markers.c xmpp-vala/src/module/xep/0334_message_processing_hints.c xmpp-vala/src/module/xep/0359_unique_stable_stanza_ids.c xmpp-vala/src/module/xep/0363_http_file_upload.c xmpp-vala/src/module/xep/0368_srv_records_tls.c xmpp-vala/src/module/xep/0380_explicit_encryption.c xmpp-vala/src/module/xep/0391_jingle_encrypted_transports.c xmpp-vala/src/module/xep/0410_muc_self_ping.c xmpp-vala/src/module/xep/pixbuf_storage.c xmpp-vala/src/util.c exports/xmpp-vala.vapi exports/xmpp-vala_internal.vapi exports/xmpp-vala.h exports/xmpp-vala_internal.h 
cd /dev/shm/dino/build/xmpp-vala && /usr/bin/valac -C --header=/dev/shm/dino/build/exports/xmpp-vala.h --internal-header=/dev/shm/dino/build/exports/xmpp-vala_internal.h --vapi=/dev/shm/dino/build/exports/xmpp-vala.vapi --internal-vapi=/dev/shm/dino/build/exports/xmpp-vala_internal.vapi -b /dev/shm/dino/xmpp-vala -d /dev/shm/dino/build/xmpp-vala --pkg=gdk-pixbuf-2.0 --pkg=gee-0.8 --pkg=gio-2.0 --pkg=glib-2.0 --pkg=gobject-2.0 --pkg=icu-uc --define=ALPN_SUPPORT --vapidir=/dev/shm/dino/xmpp-vala/vapi -g --target-glib=2.38 /dev/shm/dino/xmpp-vala/src/core/namespace_state.vala /dev/shm/dino/xmpp-vala/src/core/stanza_attribute.vala /dev/shm/dino/xmpp-vala/src/core/stanza_node.vala /dev/shm/dino/xmpp-vala/src/core/stanza_reader.vala /dev/shm/dino/xmpp-vala/src/core/stanza_writer.vala /dev/shm/dino/xmpp-vala/src/core/xmpp_log.vala /dev/shm/dino/xmpp-vala/src/core/xmpp_stream.vala /dev/shm/dino/xmpp-vala/src/module/bind.vala /dev/shm/dino/xmpp-vala/src/module/bookmarks_provider.vala /dev/shm/dino/xmpp-vala/src/module/conference.vala /dev/shm/dino/xmpp-vala/src/module/iq/module.vala /dev/shm/dino/xmpp-vala/src/module/iq/stanza.vala /dev/shm/dino/xmpp-vala/src/module/jid.vala /dev/shm/dino/xmpp-vala/src/module/message/module.vala /dev/shm/dino/xmpp-vala/src/module/message/stanza.vala /dev/shm/dino/xmpp-vala/src/module/presence/flag.vala /dev/shm/dino/xmpp-vala/src/module/presence/module.vala /dev/shm/dino/xmpp-vala/src/module/presence/stanza.vala /dev/shm/dino/xmpp-vala/src/module/roster/flag.vala /dev/shm/dino/xmpp-vala/src/module/roster/item.vala /dev/shm/dino/xmpp-vala/src/module/roster/module.vala /dev/shm/dino/xmpp-vala/src/module/roster/versioning_module.vala /dev/shm/dino/xmpp-vala/src/module/sasl.vala /dev/shm/dino/xmpp-vala/src/module/session.vala /dev/shm/dino/xmpp-vala/src/module/stanza.vala /dev/shm/dino/xmpp-vala/src/module/stanza_error.vala /dev/shm/dino/xmpp-vala/src/module/stream_error.vala /dev/shm/dino/xmpp-vala/src/module/tls.vala /dev/shm/dino/xmpp-vala/src/module/util.vala /dev/shm/dino/xmpp-vala/src/module/xep/0048_bookmarks.vala /dev/shm/dino/xmpp-vala/src/module/xep/0048_conference.vala /dev/shm/dino/xmpp-vala/src/module/xep/0402_bookmarks2.vala /dev/shm/dino/xmpp-vala/src/module/xep/0004_data_forms.vala /dev/shm/dino/xmpp-vala/src/module/xep/0030_service_discovery/flag.vala /dev/shm/dino/xmpp-vala/src/module/xep/0030_service_discovery/identity.vala /dev/shm/dino/xmpp-vala/src/module/xep/0030_service_discovery/info_result.vala /dev/shm/dino/xmpp-vala/src/module/xep/0030_service_discovery/item.vala /dev/shm/dino/xmpp-vala/src/module/xep/0030_service_discovery/items_result.vala /dev/shm/dino/xmpp-vala/src/module/xep/0030_service_discovery/module.vala /dev/shm/dino/xmpp-vala/src/module/xep/0045_muc/flag.vala /dev/shm/dino/xmpp-vala/src/module/xep/0045_muc/module.vala /dev/shm/dino/xmpp-vala/src/module/xep/0045_muc/status_code.vala /dev/shm/dino/xmpp-vala/src/module/xep/0047_in_band_bytestreams.vala /dev/shm/dino/xmpp-vala/src/module/xep/0049_private_xml_storage.vala /dev/shm/dino/xmpp-vala/src/module/xep/0054_vcard/module.vala /dev/shm/dino/xmpp-vala/src/module/xep/0060_pubsub.vala /dev/shm/dino/xmpp-vala/src/module/xep/0065_socks5_bytestreams.vala /dev/shm/dino/xmpp-vala/src/module/xep/0066_out_of_band_data.vala /dev/shm/dino/xmpp-vala/src/module/xep/0077_in_band_registration.vala /dev/shm/dino/xmpp-vala/src/module/xep/0082_date_time_profiles.vala /dev/shm/dino/xmpp-vala/src/module/xep/0084_user_avatars.vala /dev/shm/dino/xmpp-vala/src/module/xep/0085_chat_state_notifications.vala /dev/shm/dino/xmpp-vala/src/module/xep/0115_entity_capabilities.vala /dev/shm/dino/xmpp-vala/src/module/xep/0166_jingle.vala /dev/shm/dino/xmpp-vala/src/module/xep/0184_message_delivery_receipts.vala /dev/shm/dino/xmpp-vala/src/module/xep/0191_blocking_command.vala /dev/shm/dino/xmpp-vala/src/module/xep/0198_stream_management.vala /dev/shm/dino/xmpp-vala/src/module/xep/0199_ping.vala /dev/shm/dino/xmpp-vala/src/module/xep/0203_delayed_delivery.vala /dev/shm/dino/xmpp-vala/src/module/xep/0234_jingle_file_transfer.vala /dev/shm/dino/xmpp-vala/src/module/xep/0260_jingle_socks5_bytestreams.vala /dev/shm/dino/xmpp-vala/src/module/xep/0261_jingle_in_band_bytestreams.vala /dev/shm/dino/xmpp-vala/src/module/xep/0280_message_carbons.vala /dev/shm/dino/xmpp-vala/src/module/xep/0308_last_message_correction.vala /dev/shm/dino/xmpp-vala/src/module/xep/0313_message_archive_management.vala /dev/shm/dino/xmpp-vala/src/module/xep/0333_chat_markers.vala /dev/shm/dino/xmpp-vala/src/module/xep/0334_message_processing_hints.vala /dev/shm/dino/xmpp-vala/src/module/xep/0359_unique_stable_stanza_ids.vala /dev/shm/dino/xmpp-vala/src/module/xep/0363_http_file_upload.vala /dev/shm/dino/xmpp-vala/src/module/xep/0368_srv_records_tls.vala /dev/shm/dino/xmpp-vala/src/module/xep/0380_explicit_encryption.vala /dev/shm/dino/xmpp-vala/src/module/xep/0391_jingle_encrypted_transports.vala /dev/shm/dino/xmpp-vala/src/module/xep/0410_muc_self_ping.vala /dev/shm/dino/xmpp-vala/src/module/xep/pixbuf_storage.vala /dev/shm/dino/xmpp-vala/src/util.vala /dev/shm/dino/xmpp-vala/src/glib_fixes.vapi
/dev/shm/dino/xmpp-vala/src/module/xep/0166_jingle.vala:652.21-652.100: warning: unhandled error `Xmpp.Xep.Jingle.Error'
        transport = new_transport.create_transport_parameters(stream, local_full_jid, peer_full_jid);
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/dev/shm/dino/xmpp-vala/src/module/xep/0368_srv_records_tls.vala:43.13-43.51: error: The name `set_advertised_protocols' does not exist in the context of `GLib.TlsConnection?'
            tls_connection.set_advertised_protocols(new string[]{"xmpp-client"});
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Compilation failed: 1 error(s), 1 warning(s)
```

**Comment by paulmenzel** [comment]
Updating to Vala 0.50.1 fixes the issue. It’d be great if a Vala version check could be added.

**Comment by mar-v-in** [comment]
The corresponding method was added to GLib 2.60 and vapi repository 0.43.4 (via commit GNOME/vala@dd53026534a7cd8d7b049a1b09266684b6a47d2e). Before trying to build with ALPN support, we check that you have at least GLib 2.60. The first GNOME core distribution to include GLib 2.60 is 3.31.92 which also included vapi repository 0.43.92, thus every GNOME core distribution including GLib 2.60 also should be using a vapi repository that is new enough to include ALPN support. Unfortunately there is no easy way to detect the version of the vapi file at build time, so we must rely on distributions to properly follow the GNOME core distribution. What distribution are you using?

Given that most distributions put the vapi repository and vala compiler in the same package or make vala compiler depend on the vapi repository of the same version, we could add a check vor a vala compiler version > 0.43.4 before enabling ALPN so that others don't face the same issue. This would mean people with systems not following GNOME core distribution would end up without ALPN support even when their system would actually support it. But I guess that's probably the best we can do for people using such distributions :shrug:

**Comment by paulmenzel** [comment]
Thank you very much for the information about Vala. We are using our own distribution [MarIuX](http://mariux64.molgen.mpg.de/).

With the new information about how Vala works, for us no further improvement is needed.
